src/Commands/CommercePrintfulCommands.php

```diff
@@ -0,0 +1,178 @@
+ *   - http://cgit.drupalcode.org/devel/tree/src/Commands/DevelCommands.php
+ *   - http://cgit.drupalcode.org/devel/tree/drush.services.yml
+ */
+class CommercePrintfulCommands extends DrushCommands {
+
+  /**
+   * Drupal\commerce_printful\Service\Printful definition.
+   *
+   * @var \Drupal\commerce_printful\Service\Printful
+   */
+  protected $pf;
+
+  /**
+   * Drupal\commerce_printful\Service\ProductIntegrator definition.
+   *
+   * @var \Drupal\commerce_printful\Service\ProductIntegrator
+   */
+  protected $integrator;
+
+  /**
+   * Drupal\Core\Entity\EntityTypeBundleInfo definition.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeBundleInfo
+   */
+  protected $bundleInfo;
+
+  /**
+   * Drupal\Core\Config\ConfigFactoryInterface definition.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $printfulConfig;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(
+    Printful $printful,
+    ProductIntegrator $product_integrator,
+    EntityTypeBundleInfoInterface $bundle_info,
+    ConfigFactoryInterface $config_factory
+  ) {
+    $this->pf = $printful;
+    $this->integrator = $product_integrator;
+    $this->bundleInfo = $bundle_info;
+    $this->printfulConfig = $config_factory;
+  }
+
+
+    /**
+     * @return \Drupal\Core\Config\ConfigFactoryInterface
+     */
+  public function getConfig() {
+      return $this->printfulConfig;
+  }
+
+  /**
+   * Test the Printful API connection.
+   *
+   *
+   * @command printful:test
+   * @aliases pt,printful-test
+   */
+  public function test() {
+    try {
+      $store_data = $this->pf->getStoreInfo();
+      if ($store_data['code'] == 200 && isset($store_data['result'])) {
+        $d = print_r($store_data['result'], TRUE);
+        $this->output()->writeln(dt('Store Info:'));
+        $this->output()->writeln($d);
+      }
+      else {
+        $this->output()->writeln(dt('No Store Info found.'));
+      }
+    }
+    catch (PrintfulException $e) {
+      dt($e->getMessage);
+    }
+    try {
+      $products = $this->pf->syncProducts();
+      if ($products['code'] == 200 && isset($products['result'])) {
+        $p = print_r($products['result'], TRUE);
+        $this->output()->writeln(dt('Products:'));
+        $this->output()->writeln($p);
+      }
+      else {
+        $this->output()->writeln(dt('No products found.'));
+      }
+    }
+    catch (PrintfulException $e) {
+      dt($e->getMessage);
+    }
+  }
+
+  /**
+   * Sync Printful products to Drupal Commerce products.
+   *
+   *
+   * @command printful:sync-products
+   * @param $bundle A string machine_name of the bundle to sync.
+   * @usage drush printful:sync-products
+   * @aliases psp,printful-sync-products
+   */
+  public function syncProducts($bundle, $update) {
+
+    $batch = PrintfulSyncBatch::getBatch([
+      'product_bundle' => $bundle,
+      'update' => $update,
+    ]);
+    batch_set($batch);
+
+    drush_backend_batch_process();
```

Comment:
Yes, there's that drush function.

Maybe worth adding input validation and a try - catch block to catch a `PrintfulException`?

Just FYI, in VBO I did it like this (I don't remember why but had my reasons), using `drush_backend_batch_process` is OK though:

```    // Populate entity list.
    $context = [];
    do {
      $context['finished'] = 1;
      $context['message'] = '';
      ViewsBulkOperationsBatch::getList($vbo_data, $context);
      if (!empty($context['message'])) {
        $this->logger->info($context['message']);
      }
    } while ($context['finished'] < 1);
    $vbo_data = $context['results'];

    $this->timer($options['verbose'], 'list');

    // Execute the selected action.
    $context = [];
    do {
      $context['finished'] = 1;
      $context['message'] = '';
      ViewsBulkOperationsBatch::operation($vbo_data, $context);
      if (!empty($context['message'])) {
        $this->logger->info($context['message']);
      }
    } while ($context['finished'] < 1);
```

Reply:
Yeah, I did look at your VBO code.  I found that core removed the type hint for a bunch of stuff due to this very issue https://www.drupal.org/node/2664290.  I wonder if you ran into the same thing and just created your own work around?  Personal feeling is, if it's good enough for core, I am leaning toward leaving as-is.